# *N. Gonorrhoeae* SNP mining with minSNPs
Data source: [Pathogen**watch**](https://pathogen.watch/genomes/all?organismId=485)
- Assembled contigs: [https://pathogenwatch-public.ams3.cdn.digitaloceanspaces.com/Neisseria%20gonorrhoeae__fastas.zip](https://pathogenwatch-public.ams3.cdn.digitaloceanspaces.com/Neisseria%20gonorrhoeae__fastas.zip)
The zip file containing all the assembled contigs need to be unzipped with `7zz e -aou`, duplicated assembly can be found in the dataset, when duplicated, "_1" is added to the file name.
- Metadatas: 
    - [https://pathogenwatch-public.ams3.cdn.digitaloceanspaces.com/Neisseria%20gonorrhoeae__metadata.csv.gz](https://pathogenwatch-public.ams3.cdn.digitaloceanspaces.com/Neisseria%20gonorrhoeae__metadata.csv.gz)
    - [https://pathogenwatch-public.ams3.cdn.digitaloceanspaces.com/Neisseria%20gonorrhoeae__mlst.csv.gz](https://pathogenwatch-public.ams3.cdn.digitaloceanspaces.com/Neisseria%20gonorrhoeae__mlst.csv.gz)
    - [https://pathogenwatch-public.ams3.cdn.digitaloceanspaces.com/Neisseria%20gonorrhoeae__paarsnp.csv.gz](https://pathogenwatch-public.ams3.cdn.digitaloceanspaces.com/Neisseria%20gonorrhoeae__paarsnp.csv.gz)

# Combining the metadata

# Generating the SNP matrix

# Preprocessing the SNP matrix

# Running minSNPs

